subt/simulation.py

```diff
@@ -0,0 +1,159 @@
+
+_saved_print = print
+
+def verbose(ok):
+    print(ok)
+    if ok:
+        globals()['print'] = _saved_print
+    else:
+        globals()['print'] = lambda *args: None
+
+
+def log(*args):
+    up = inspect.currentframe().f_back
+    caller = up.f_code.co_name
+    caller_class = up.f_locals['self'].__class__
+    color_name = getattr(caller_class, caller).color
+    print(stylize(f"{caller_class.__name__}.{caller}", fg(color_name)), *args)
+
+
+def clip(min_value, value, max_value):
+    return max(min(value, max_value), min_value)
+
+
+class Color:
+    def __getattr__(self, color_name):
+        def a(method):
+            method.color = color_name
+            return method
+        return a
+
+
+color = Color()
+
+
+class SimLogger:
+
+    def __init__(self):
+        self.lock = RLock()
+        self.start_time = datetime.datetime.now(datetime.timezone.utc)
+        self.names = ["sys"]
+
+    def register(self, name):
+        self.names.append(name)
+        return len(self.names)-1
+
+    @color.dodger_blue_1
+    def write(self, stream_id, data, dt=None):
+        #log(self.names[stream_id], data)
+        dt = datetime.datetime.now(datetime.timezone.utc) - self.start_time
+        return dt
+
+
+class Simulation:
+
+    def __init__(self, bus, end_condition):
+        bus.register('scan', 'rot', 'orientation', 'sim_time_sec', 'origin', 'pose2d', 'acc', 'artf')
+        self.bus = bus
+        self.end_condition = end_condition
+        self._handlers = collections.defaultdict(lambda: self.on_default)
+        self.set_handler('request_origin', self.on_request_origin)
+        self.set_handler('desired_speed', self.on_desired_speed)
+        self.set_handler('pose2d', self.on_pose)
+        self.set_handler('pose3d', self.on_pose)
+        self.origin = [7.0, 3.0, 0.0]
+        self.xyz = [0.0, 0.0, 0.0]
+        self.orientation = quaternion.identity()
+        self.speed_forward = 0.0
+        self.speed_angular = 0.0
+        self.time = datetime.timedelta()
+
+    def start(self):
+        self.thread = Thread(target=self.main)
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout)
+
+    @color.green
+    def main(self):
+        log("started")
+        self.on_request_origin(datetime.timedelta(), "request_origin", True)
+
+        for _ in range(500):
+            packet = self.bus.listen()
+            self.handle_incomming(*packet)
+            if self.end_condition(self):
+                return
+
+    def set_handler(self, channel, handler):
+        self._handlers[channel] = handler
+
+    def handle_incomming(self, dt, channel, data):
+        self._handlers[channel](dt, channel, data)
+
+    @color.red
+    def on_default(self, dt, channel, data):
+        log(dt, channel, data)
+
+    @color.gold_1
+    def on_request_origin(self, dt, channel, data):
+        log(dt, channel, data)
+        # start sending origin data until robot arrives within 0.3m from (0,0,0)
```

Comment:
Where are the 0.3 and (0, 0, 2) parts in the code?

Reply:
That is actually forgotten TODO. It is how ros_proxy_node.cc behaves. It is here

https://github.com/robotika/osgar/blob/be3010c2b0a268721b990cdb07cc5dc823e73dfa/subt/ros/proxy/ros_proxy_node.cc#L416-L421

But the current code in the simulation sends the origin only once (actually twice - first to get robot name, second to get the origin). The SubT application does not require the origin to be sent more than once. Sending it many times probably should help remove any positional error accumulated in the start area. I have reworded the comment to acknowledge the difference for now.